blocks/fileio/include/gnuradio-4.0/fileio/BasicFileIo.hpp

```diff
@@ -0,0 +1,307 @@
+
+    std::sort(matchingFiles.begin(), matchingFiles.end());
+    return matchingFiles;
+}
+
+[[nodiscard]] inline std::uintmax_t getFileSize(const std::filesystem::path& filePath) {
+    if (!std::filesystem::exists(filePath)) {
+        throw std::runtime_error(fmt::format("file '{}' does not exist.", filePath.string()));
+    }
+    return std::filesystem::file_size(filePath);
+}
+
+[[maybe_unused]] inline std::vector<std::string> deleteFilesContaining(const std::string& fileName) {
+    std::filesystem::path filePath(fileName);
+    if (!std::filesystem::exists(filePath.parent_path())) {
+        return {};
+    }
+
+    std::vector<std::string> deletedFiles;
+    for (const auto& entry : std::filesystem::directory_iterator(filePath.parent_path())) {
+        if (entry.is_regular_file() && entry.path().string().find(filePath.filename().string()) != std::string::npos) {
+            deletedFiles.push_back(entry.path().string());
+            std::filesystem::remove(entry.path());
+        }
+    }
+
+    return deletedFiles;
+}
+
+} // namespace detail
+
+enum class FileMode { override, append, multi };
+
+template<typename T>
+struct BasicFileSink : public gr::Block<BasicFileSink<T>> {
+    using Description = Doc<R""(A sink block for writing a stream to a binary file.
+The file can be played back using a 'BasicFileSource' or read by any program that supports binary files (e.g. Python, C, C++, MATLAB).
+For complex types, the binary file contains [float, double]s in IQIQIQ order. No metadata is included with the binary data.)
+Important: this implementation assumes a host-order, CPU architecture specific byte order!)"">;
+    template<typename U, gr::meta::fixed_string description = "", typename... Arguments>
+    using A = gr::Annotated<U, description, Arguments...>; // optional shortening
+
+    PortIn<T> in;
+
+    A<std::string, "file name", Doc<"base filename, prefixed if ">, Visible>              file_name;
+    FileMode                                                                              _mode              = FileMode::override;
+    A<std::string, "mode", Doc<"mode: \"override\", \"append\", \"multi\"">, Visible>     mode               = std::string(magic_enum::enum_name(_mode));
+    A<gr::Size_t, "max bytes per file", Doc<"max bytes per file, 0: infinite ">, Visible> max_bytes_per_file = 0U;
+
+    std::size_t   _totalBytesWritten{0UZ};
+    std::size_t   _totalBytesWrittenFile{0UZ};
+    std::ofstream _file;
+    std::size_t   _fileCounter{0UZ};
+    std::string   _actualFileName;
+
+    void settingsChanged(const property_map& /*oldSettings*/, const property_map& /*newSettings*/) {
+        _mode = magic_enum::enum_cast<FileMode>(mode, magic_enum::case_insensitive).value_or(_mode);
```

Comment:
Yes, correct. If it's thrown then an exception may be thrown that is intercepted by `Block<T>` (as all exceptions are in user-defined functions), converted to an error message and forwarded via the scheduler back to the user.

Reply:
N.B. nothing needs to be done here because the function already has a try-catch block in the calling function e.g.
https://github.com/fair-acc/gnuradio4/blob/f00cba2c81422f143c9d48552921ec63bbaef652/core/include/gnuradio-4.0/Block.hpp#L921-L935
and
https://github.com/fair-acc/gnuradio4/blob/f00cba2c81422f143c9d48552921ec63bbaef652/core/include/gnuradio-4.0/Block.hpp#L508-L524

Reply:
My point was not about throwing exceptions. I believe that `value_or()` can't throw any exception in this case.

My point was that this piece of code is converting the settings provided string to a an enum value, or keeping the current `_mode` if the conversion results in an error (that's what `value_or()` is doing here).

Thus, it appears that the behaviour of this code is that when new settings are given, the `_mode` is changed to the one indicated by the new settings *if the new `mode` can be parsed* and otherwise the parse error is silently ignored and the old `_mode` is kept. This is what I find potentially surprising for the user. It would be better to error out (for instance throw an exception) if the `mode` cannot be parsed.